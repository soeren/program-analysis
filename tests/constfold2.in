// Test for constant folding
// - Check intra procedural folding

program

int x;

int foldme(int a, int b){
    int y;
    int z;
    int r;

    y := 23;
    z := 42;
    r := y * z; // should be folded to 966

    if(r > a){ // a not const -> folding to if(966 > a)
        y := b; // b not const -> stay the same
    } else {
        z := b; // b not const -> stay the same
    }

    r := r - 666; // r is const -> folding to 300

    while(b > y | b > z){
        r := r -1; // r not const anymore -> no folding
    }

    return r; // r not const anymore -> no folding
}

int givefive(int a){
    int q;
    int w;

    q:=1;
    w:=3;

    return q*w; // q and w const -> folding to 3
}

x := foldme(1,2) + givefive(0);
write(x);

end
