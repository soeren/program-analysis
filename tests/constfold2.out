program
int x;
int foldme(int a,int b) // 3
{
     int y;
     int z;
     int r;
     y := 23; // 4
     z := 42; // 5
     r := 966; // 6
     if(966 > a) // 7
     {
          y := b; // 8
      } else
     {
          z := b; // 9
      }
     r := 300; // 10
     while(b > y | b > z) // 11
     {
          r := r - 1; // 12
      }
     return r; // 13
 }
int givefive(int a) // 14
{
     int q;
     int w;
     q := 1; // 15
     w := 3; // 16
     return 3; // 17
 }
x := foldme(1,2) + 3; // 1
write(x); // 2
end
